deepmd/pt/model/descriptor/repformer_layer.py

```diff
@@ -431,165 +872,74 @@ def _update_g1_conv(
         nlist_mask: torch.Tensor,
         sw: torch.Tensor,
     ) -> torch.Tensor:
+        """
+        Calculate the convolution update for atomic invariant rep.
+
+        Parameters
+        ----------
+        gg1
+            Neighbor-wise atomic invariant rep, with shape nf x nloc x nnei x ng1.
+        g2
+            Pair invariant rep, with shape nf x nloc x nnei x ng2.
+        nlist_mask
+            Neighbor list mask, where zero means no neighbor, with shape nf x nloc x nnei.
+        sw
+            The switch function, which equals 1 within the rcut_smth range, smoothly decays from 1 to 0 between rcut_smth and rcut,
+            and remains 0 beyond rcut, with shape nf x nloc x nnei.
+        """
         assert self.proj_g1g2 is not None
-        nb, nloc, nnei, _ = g2.shape
+        nf, nloc, nnei, _ = g2.shape
         ng1 = gg1.shape[-1]
         ng2 = g2.shape[-1]
-        # gg1  : nb x nloc x nnei x ng2
-        gg1 = self.proj_g1g2(gg1).view(nb, nloc, nnei, ng2)
-        # nb x nloc x nnei x ng2
+        # gg1  : nf x nloc x nnei x ng2
+        gg1 = self.proj_g1g2(gg1).view(nf, nloc, nnei, ng2)
+        # nf x nloc x nnei x ng2
         gg1 = _apply_nlist_mask(gg1, nlist_mask)
         if not self.smooth:
             # normalized by number of neighbors, not smooth
-            # nb x nloc x 1
-            invnnei = 1.0 / (self.epsilon + torch.sum(nlist_mask, dim=-1)).unsqueeze(-1)
+            # nf x nloc x 1
+            # must use type_as here to convert bool to float, otherwise there will be numerical difference from numpy
+            invnnei = 1.0 / (
+                self.epsilon + torch.sum(nlist_mask.type_as(gg1), dim=-1)
+            ).unsqueeze(-1)
         else:
             gg1 = _apply_switch(gg1, sw)
             invnnei = (1.0 / float(nnei)) * torch.ones(
-                (nb, nloc, 1), dtype=env.GLOBAL_PT_FLOAT_PRECISION, device=gg1.device
+                (nf, nloc, 1), dtype=gg1.dtype, device=gg1.device
             )
-        # nb x nloc x ng2
+        # nf x nloc x ng2
         g1_11 = torch.sum(g2 * gg1, dim=2) * invnnei
         return g1_11
 
-    def _cal_h2g2(
```

Comment:
I know self is not used but do not move the position of cal_h2g2. you may make it a static method if you want